shared/pubsub.js

```diff
@@ -0,0 +1,397 @@
+}
+
+export function createRequest (type: $Values<typeof REQUEST_TYPE>, data: JSONObject): string {
+  // Had to use Object.assign() instead of object spreading to make Flow happy.
+  return JSON.stringify(Object.assign({ type }, data))
+}
+
+// ====== Event Handlers ====== //
+
+const defaultHandlers = {
+  close (event: CloseEvent) {
+    if (this.socket) {
+      // Remove event listeners to avoid memory leaks.
+      for (const name of eventNames) {
+        this.socket.removeEventListener(name, this.listeners[name])
+      }
+    }
+    this.socket = null
+
+    if (this.connectionTimeoutID) {
+      clearTimeout(this.connectionTimeoutID)
+      this.connectionTimeoutID = undefined
+    }
+    if (this.isReconnecting) {
+      this.failedAttempts++
+    }
+    if (event.reason === 'timeout' && this.options.reconnectOnTimeout) {
+      this.reconnectionDelayID = setTimeout(
+        this.connect,
+        this.getNextRandomReconnectionDelay()
+      )
+    }
+  },
+
+  // Do not manually close the socket here as it will be done automatically.
+  error (event: Event) {
+  },
+
+  message (event: MessageEvent) {
+    const { data } = event
+
+    if (typeof data !== 'string') {
+      return
+    }
+
+    let msg: Message = { type: '' }
+    try {
+      msg = messageParser(data)
+    } catch (error) {
+      // TODO: place us in unrecoverable state (see state.js error handling TODOs)
+      console.error(`[pubsub] Malformed message: ${error.message}`)
+      return
+    }
+    const handler = this.messageHandlers[msg.type]
+
+    if (handler) {
+      handler.call(this, msg)
+    } else {
+      throw new Error(`Unhandled message type: ${msg.type}`)
+    }
+  },
+
+  open (event) {
+    this.pendingSubscriptionSet.forEach(this.sub)
+    this.pendingUnsubscriptionSet.forEach(this.unsub)
+
+    if (this.isReconnecting) {
+      console.log('[pubsub] Connection re-established')
+      this.failedAttempts = 0
+      this.isReconnecting = false
+    }
+    if (this.connectionTimeoutID) {
+      clearTimeout(this.connectionTimeoutID)
+      this.connectionTimeoutID = undefined
+    }
+  }
+}
+
+// ====== Message Handlers ====== //
+
+const defaultMessageHandlers = {
+  // PUB can be used to send ephemeral messages outside of any contract log.
+  [NOTIFICATION_TYPE.PUB] (msg) {
+    console.debug(`[pubsub] Ignoring ${msg.type} message:`, msg.data)
+  },
+  [NOTIFICATION_TYPE.SUB] (msg) {
+    console.debug(`[pubsub] Ignoring ${msg.type} message:`, msg.data)
+  },
+  [NOTIFICATION_TYPE.UNSUB] (msg) {
+    console.debug(`[pubsub] Ignoring ${msg.type} message:`, msg.data)
+  },
+  [RESPONSE_TYPE.ERROR] ({ data: { type, contractID } }) {
+    console.log(`[pubsub] Received ERROR response for ${type} request to ${contractID}`)
+  },
+  [RESPONSE_TYPE.SUCCESS] ({ data: { type, contractID } }) {
+    const client = this
+    switch (type) {
+      case REQUEST_TYPE.SUB: {
+        console.log(`[pubsub] Subscribed to ${contractID}`)
```

Comment:
Two questions:

1. Do the sockets on this side (the frontend) share the same socket ID as the one on the backend? E.g. can you access it through `this` or `client`?
2. If so, can you log that here?
3. If not, can it be added? This will help with debugging and being able to match console logs between the frontend and backend when there are multiple clients/sockets open.

Reply:
1. No, these sockets have currently no ID at all. The client doesn't know which ID the server has chosen, so it's currently not accessible. I thought initially that such an ID was not necessary since a pubsub client manages a single raw socket at a time.
2. This would be very useful in the backend console when running unit tests. Not so much in the browser console though, since there is only one socket per frontend at a time.
3. It can be added but there is a few different ways:
- By implementing an equivalent to the [primus.id(callback)](https://github.com/primus/primus/#primusidcallback) method, which fetches the relevant socket ID from the server. The downside being the cost of the additional request-response cycle. Besides, Primus ID's are `uuid-v4` ID's, not plain integers starting from zero.
- By letting the front-end generate an ID and send it to server, e.g. as query parameter in the WebSocket URL.

Another idea to ease debugging is to log the current username along or instead the socket ID, or "anonymous" if the connection is not authenticated. But I've noticed that currently, only opaque socket unique IDs or identity contract IDs are ever logged in the backend. If there any specific reason about this?

Reply:
> Another idea to ease debugging is to log the current username along or instead the socket ID, or "anonymous" if the connection is not authenticated.

Something alone these lines sounds like the most reasonable way to go forward to me. Either logging the username or, more likely, the pubkey hash (which we don't have yet). In other words, something the backend already has to avoid the request/response cycle.

> But I've noticed that currently, only opaque socket unique IDs or identity contract IDs are ever logged in the backend. If there any specific reason about this?

If I understand your question (and I might be misunderstanding), the reason is that we don't have real authentication built-in yet.

Reply:
After investigation, I think passing the username to the server would not work as the same connection seems to be reused when a user logs out then logs in again as another user without reloading the page, which happens a lot in cypress tests.

That behavior should be fixed, but for now I guess it's ok to use a dummy devmode-only random socket ID with an explicit comment.